trillian/integration/ct_integration.go

```diff
@@ -589,6 +593,178 @@ func RunCTIntegrationForLog(cfg *configpb.LogConfig, servers, metricsServers, te
 	return nil
 }
 
+// RunCTLifecycleForLog does a simple log lifecycle test. The log
+// is assumed to be newly created when this test runs. A random number of
+// entries are then submitted to build up a queue. The log is set to
+// DRAINING state and the test checks that all the entries are integrated
+// into the tree and we can verify a consistency proof to the latest entry.
+func RunCTLifecycleForLog(cfg *configpb.LogConfig, servers, metricsServers, adminServer string, testdir string, mmd time.Duration, stats *logStats) error {
+	// Retrieve the test data.
+	testDir := "../testdata"
+	caChain, err := GetChain(testDir, "int-ca.cert")
+	if err != nil {
+		return err
+	}
+	leafChain, err := GetChain(testDir, "leaf01.chain")
+	if err != nil {
+		return err
+	}
+	signer, err := MakeSigner(testDir)
+	if err != nil {
+		return err
+	}
+	leafCert, err := x509.ParseCertificate(leafChain[0].Data)
+	if err != nil {
+		return err
+	}
+	caCert, err := x509.ParseCertificate(caChain[0].Data)
+	if err != nil {
+		return err
+	}
+	ctx := context.Background()
+	pool, err := NewRandomPool(servers, cfg.PublicKey, cfg.Prefix)
+	if err != nil {
+		return fmt.Errorf("failed to create pool: %v", err)
+	}
+	t := testInfo{
+		prefix:         cfg.Prefix,
+		cfg:            cfg,
+		metricsServers: metricsServers,
+		adminServer:    adminServer,
+		stats:          stats,
+		pool:           pool,
+	}
+
+	if err := t.checkStats(); err != nil {
+		return fmt.Errorf("unexpected stats check: %v", err)
+	}
+
+	// Stage 0: get accepted roots, which should just be the fake CA.
+	roots, err := t.client().GetAcceptedRoots(ctx)
+	t.stats.done(ctfe.GetRootsName, 200)
+	if err != nil {
+		return fmt.Errorf("got GetAcceptedRoots()=(nil,%v); want (_,nil)", err)
+	}
+	if len(roots) != 1 {
+		return fmt.Errorf("len(GetAcceptedRoots())=%d; want 1", len(roots))
+	}
+
+	// Stage 1: get the STH, which should be empty.
+	sth0, err := t.client().GetSTH(ctx)
+	t.stats.done(ctfe.GetSTHName, 200)
+	if err != nil {
+		return fmt.Errorf("got GetSTH()=(nil,%v); want (_,nil)", err)
+	}
+	if sth0.Version != 0 {
+		return fmt.Errorf("sth.Version=%v; want V1(0)", sth0.Version)
+	}
+	if sth0.TreeSize != 0 {
+		return fmt.Errorf("sth.TreeSize=%d; want 0", sth0.TreeSize)
+	}
+	fmt.Printf("%s: Got STH(time=%q, size=%d): roothash=%x\n", t.prefix, timeFromMS(sth0.Timestamp), sth0.TreeSize, sth0.SHA256RootHash)
+
+	// Stage 2: add certificates 2, 3, 4, 5,...N, for some random N with
+	// at least 2000 so the queue builds up a bit.
+	atLeast := 2000
+	count := atLeast + rand.Intn(3000-atLeast)
+	fmt.Printf("%s: Starting upload of %d certificates ....\n", t.prefix, count)
+	for i := 1; i <= count; i++ {
+		chain, err := makeCertChain(leafChain, leafCert, caCert, signer, time.Now().Add(24*time.Hour))
+		if err != nil {
+			return err
+		}
+		_, err = t.client().AddChain(ctx, chain)
+		t.stats.done(ctfe.AddChainName, 200)
```

Comment:
should this be after the err check? (what's the 200 mean - success or something else?)

Reply:
No. It means we expect a 200 for that request. If we get something else it will fail the stats check.

Reply:
could we rename the method or something to make that clearer? 
it reads very much like "update stats, a request was done and had code X"

Reply:
I've renamed it.